### ISSUE TITLE
Bug: [explicit-module-boundary-types] `allowDirectConstAssertionInArrowFunctions` should not error when also using `satisfies`

### ISSUE THREAD

**Issue by jleider** [typescript-eslint/typescript-eslint#10231]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.3&fileType=.ts&code=FAFwngDgpgBAYgewTAvDA3sGMBmSBcMAziAE4CWAdgOYA0WMARgIamGUCuAto1KfQF8A3MGBQAHhASkQMAMYJKJGM1QwAFAEpUAPg2ZseBIQBEcAPKWT9bCzYwAjACYAzIOZF5ikpqEwA9P4wADwAtKEwACoAFuSeAO7SANaeHjAS0HIgUAAmAHSiGdKyCkqyjGpauvoMRqYWVjZMrITObsACKp6lykTMIHE45FCeiAi%2BAUFhETFxMCDRpAjxqZTppEukBUA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6RAD2MbMroFsA9gBMEiWgCNBsJsICG0QrSKlUGANrhsORNGiDokADRbtwKHPjxBAdwAilaIjL4AwoKbJ8AQWTJd%2BJQeAJJM3nq2AGIyLkGe6GD4cIhgAL6mYAC6WumpQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type Foo = {
  foo: string,
  bar: number,
};

export const a = () => ({
  foo: "FOOO",
  bar: 123,
}as const); // <-- This works as expected.

export const b = () => ({
  foo: "FOOO",
  bar: 123,
} as const satisfies Foo); // <-- This throws an error.
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": [
      "error",
      { "allowDirectConstAssertionInArrowFunctions": true }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

This concerns both `explicit-module-boundary-types` and `explicit-function-return-type` rules.

I have a situation where I want to use both `as const` followed by `satisfies`. With the rule currently as written as soon as you add a `satisfies` after an `as const` the rule will error. Leaving off the `satisfies` works as expected. 

In this scenario, the `satisfies Foo` provides extra type safety ensuring that `a satisfies Foo` but still keeping the ability to have explicit type literals for `foo: "FOOO"` and `bar: 123`. 

### Actual Result

The rule should not throw an error when using `as const satisfies Foo`

**Comment by kirkwaiblinger** [comment]
I think this request makes sense, +1. Noting though that this is related to, but not the same as, #9391.

**Comment by jleider** [comment]
> I think this request makes sense, +1. Noting though that this is related to, but not the same as, #9391.

I did see that one but from what I could tell the OP of that request wanted the `satisfies` to be sufficient for the return type rules whereas I still want the `as const` to apply even with `satisfies`.

**Comment by bradzacher** [comment]
My question would be - why are you doing `as const sarisfies T` instead of just annotating the return type?

What do you gain by doing `as const satisfies`?

**Comment by jleider** [comment]
> My question would be - why are you doing `as const sarisfies T` instead of just annotating the return type?
> 
> What do you gain by doing `as const satisfies`?

`as const satisfies T` is slightly different than explicitly annotating the return type. For example, using the same types as in the playground linked above, I've created a simple example as to why it's useful to use `as const satisfies`. Mainly its useful when you want to constrain the `as const` to some kind of interface while still maintaining the literal types created by the `as const`. It's kind of like the best of both worlds. Without the two in conjunction, it would be tedious to hand write every type explicitly to narrow the return type with type literals.

https://tsplay.dev/w1DXAW

```typescript
type Foo = {
  foo: string,
  bar: number,
};

export const a = () => ({
  foo: "FOOO",
  bar: 123,
} as const satisfies Foo); // <-- Ensures that the object returned satisfies the Foo type

/* const a: () => {
    readonly foo: "FOOO";
    readonly bar: 123;
} */

export const b = (): Foo => ({ // <-- this type annotation ignores the `as const` and returns Foo only
  foo: "FOOO",
  bar: 123,
} as const);

/* const b: () => Foo */

export const c = () => ({ 
  foo: 123, // <-- This mistake is caught by the satisfies while still maintaining the type literals when correct
  bar: "Baaar", // <-- This mistake is caught by the satisfies while still maintaining the type literals when correct
} as const satisfies Foo);

/* const c: () => {
  readonly foo: 123;
  readonly bar: "Baaar";
} */
```

**Comment by bradzacher** [comment]
Sorry, to clarify -- I fully understand the mechanical difference between the two.

I'm more asking about the usecase.
What usecase do you unlock by having a type that matches a shape, but is readonly with the specific const literal values?

**Comment by auvred** [comment]
I have used `as const satisfies` several times in a project with a complex build pipeline, where I needed to operate on packages using their names as literals.

Imagine the following:

```ts
type PackageName = '@scope/app-foo' | '@scope/app-bar' // | etc ...

export const packageSettings = {
  '@scope/app-foo': {
    option: 'foo bar'
  }
} as const satisfies Record<PackageName, { option: string }>
```

Then we can do `packageSettings['@scope/app-foo']`, but cannot `packageSettings['@scope/app-bar']`. We can also find out the names of packages with settings by simply `keyof typeof packageSettings`.

Of course we could write something like:

```ts
export const packageSettings: Record<'@scope/app-foo', { option: string }? = {
  '@scope/app-foo': {
    option: 'foo bar'
  }
}
```

But in this case we would have to write package names twice: one for type and one for value. It's easier to write only the value, and infer the type from it, while ensuring that we're still inside the allowed boundaries (`PackageName`).

**Comment by jleider** [comment]
In our general use cases, we take all the objects that use `as const satisfies` and create a tagged union of them. After that we can exhaustively discriminate the union with a switch statement to ensure we've covered every case wherever we need to. Adding to the union now creates a compile time dependency that must be met in every place that is exhaustively discriminating that union.

In particular, our largest use case for this is for routing purposes, where each object describes a route with the specific inputs, outputs, page meta data, reverse routing urls, etc. Since each page has different data, its useful to be able to know at compile time which route contains which data. We would not be able to do that as easily without the `as const satisfies`. Of course, our functions aren't simple thunks, they take params necessary to build the route data.
